### PR TITLE
Make local scheduler client thread-safe

### PR DESCRIPTION
**Pull request by raulchen**
Local scheduler client is now used in multiple threads, but it's not thread-safe.

This PR fixes this issue by adding 2 locks:
1) the 'write_mutex' is used to prevent multiple threads from writing to the connection simultaneously.
2) since the reads are always used together with writes, which makes the whole process transactional, another 'mutex' is used to make the these processes atomic.

Perf impact in single thread is negligible:
```
Benchmark: https://gist.github.com/raulchen/2c315cbd4d2387bf356adbe2a8641986#file-gistfile1-txt

With this PR:
test_put_and_get: Avg: 235.28, Min: 209.41, Max: 429.16, P50: 227.32, P90: 268.50
test_remote: Avg: 149.50, Min: 61.84, Max: 279.36, P50: 150.31, P90: 191.73

Without this PR:
test_put_and_get: Avg: 259.65, Min: 218.51, Max: 420.26, P50: 258.75, P90: 280.03
test_remote: Avg: 146.55, Min: 85.89, Max: 249.83, P50: 144.64, P90: 180.68
```

Previous discussions can be found at https://github.com/ant-tech-alliance/ray/pull/49#issuecomment-403781006.

Related issue: #2385

### CODE REVIEW

File: src/common/io.cc [hunk 1]
```diff
@@ -251,7 +251,7 @@ int write_bytes(int fd, uint8_t *cursor, size_t length) {
   return 0;
 }
 
-int write_message(int fd, int64_t type, int64_t length, uint8_t *bytes) {
+int _write_message(int fd, int64_t type, int64_t length, uint8_t *bytes) {
```
Comment (pcmoritz) [inline src/common/io.cc]: function names starting with an _ are reserved for the system/compiler, so we should name this something else (for example do_write_message)

**Comment by alexbao** [comment]
@robertnishihara @pcmoritz

File: src/common/io.cc [hunk 2]
```diff
@@ -388,8 +401,8 @@ int64_t read_vector(int fd, int64_t *type, std::vector<uint8_t> &buffer) {
 
 void write_log_message(int fd, const char *message) {
   /* Account for the \0 at the end of the string. */
-  write_message(fd, static_cast<int64_t>(CommonMessageType::LOG_MESSAGE),
```
Comment (pcmoritz) [inline src/common/io.cc]: doesn't this need to be protected by a lock too?

Comment (raulchen) [reply src/common/io.cc]: No, because this function isn't used by local scheduler client.
(Actually, I searched the code, it's only used in tests.)

Comment (pcmoritz) [reply src/common/io.cc]: Yeah, I think you are right!

**Review by pcmoritz — APPROVED** [review]
LGTM! We probably don't want to encourage too much multithreading for the local scheduler client, but it's good for it to be threadsafe!

**Comment by raulchen** [comment]
@pcmoritz Thanks for taking the time to review this PR!

I understand that it's not a good practice to use too much multithreading in a CPython program (especially when it's computation intensive). But this PR can still be useful in some aspects:
1) It fixes the bug that main thread, import thread and flush-profile-data thread are all using the local scheduler client without proper protection.
2) IO intensive Python code. One use case is that we want to run a gPRC server in ray. What the server does is simply accepting a request, forward it to an actor and wait. In this case, it makes sense to use multi-threading. Because otherwise we'll need to spawn many actors, and each of them can only handle a request at a time.
3) Java worker. Because java doesn't have the GIL, and the overhead of JVM is big.

**Comment by pcmoritz** [comment]
retest this please

**Comment by pcmoritz** [comment]
jenkins retest this please

**Pull request merged**